graylog2-server/src/main/java/org/graylog2/web/resources/ResourceFileReader.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.web.resources;
+
+import com.google.common.base.Suppliers;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.hash.Hashing;
+import com.google.common.io.Resources;
+import jakarta.inject.Inject;
+import jakarta.inject.Singleton;
+import jakarta.ws.rs.core.EntityTag;
+import org.graylog2.web.PluginAssets;
+
+import javax.annotation.Nonnull;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystemAlreadyExistsException;
+import java.nio.file.FileSystemNotFoundException;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.nio.file.attribute.FileTime;
+import java.util.Collections;
+import java.util.Date;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+@Singleton
+public class ResourceFileReader {
+    private final LoadingCache<URI, FileSystem> fileSystemCache;
+
+    @Inject
+    public ResourceFileReader() {
+        this.fileSystemCache = CacheBuilder.newBuilder()
+                .maximumSize(1024)
+                .build(new CacheLoader<>() {
+                    @Override
+                    public FileSystem load(@Nonnull URI key) throws Exception {
+                        try {
+                            return FileSystems.getFileSystem(key);
+                        } catch (FileSystemNotFoundException e) {
+                            try {
+                                return FileSystems.newFileSystem(key, Collections.emptyMap());
+                            } catch (FileSystemAlreadyExistsException f) {
+                                return FileSystems.getFileSystem(key);
+                            }
+                        }
+                    }
+                });
+    }
+
+    public record ResourceFile(java.nio.file.Path path,
+                               Supplier<byte[]> contents,
+                               Supplier<EntityTag> entityTag) {
```

Comment:
Cannot we use plain data objects instead of Suppliers? Maybe I have missed smth, but it looks like a overcomplication from a distance.

Reply:
`byte[]` should not be used in records (primitive arrays in general). `entityTag` is a `Supplier`, so it can be used lazily.